src/backend/distributed/planner/multi_explain.c

```diff
@@ -875,6 +976,289 @@ ExtractFieldJsonbDatum(Datum jsonbDoc, const char *fieldName, Datum *result)
+					   TupleDestination *defaultTupleDest,
+					   TupleDesc tupleDesc,
+					   ParamListInfo params)
+{
+	List *explainAnalyzeTaskList = NIL;
+	Task *originalTask = NULL;
+
+	/*
+	 * We cannot use multiple commands in a prepared statement, so use the old
+	 * EXPLAIN ANALYZE method for this case.
+	 */
+	if (params != NULL)
+	{
+		return originalTaskList;
+	}
+
+	foreach_ptr(originalTask, originalTaskList)
+	{
+		if (originalTask->queryCount != 1)
+		{
+			ereport(ERROR, (errmsg("cannot get EXPLAIN ANALYZE of multiple queries")));
+		}
+
+		Task *explainAnalyzeTask = copyObject(originalTask);
+		const char *queryString = TaskQueryStringForAllPlacements(explainAnalyzeTask);
+		char *wrappedQuery = WrapQueryForExplainAnalyze(queryString, tupleDesc);
+		char *fetchQuery = "SELECT worker_last_saved_explain_analyze()";
+		SetTaskQueryStringList(explainAnalyzeTask, list_make2(wrappedQuery, fetchQuery));
+
+		TupleDestination *originalTaskDest = originalTask->tupleDest ?
+											 originalTask->tupleDest :
+											 defaultTupleDest;
+
+		explainAnalyzeTask->tupleDest =
+			CreateExplainAnlyzeDestination(originalTask, originalTaskDest);
+
+		explainAnalyzeTaskList = lappend(explainAnalyzeTaskList, explainAnalyzeTask);
+	}
+
+	return explainAnalyzeTaskList;
+}
+
+
+/*
+ * WrapQueryForExplainAnalyze wraps a query into a worker_save_query_explain_analyze()
+ * call so we can fetch its explain analyze after its execution.
+ */
+static char *
+WrapQueryForExplainAnalyze(const char *queryString, TupleDesc tupleDesc)
+{
+	StringInfo columnDef = makeStringInfo();
+	for (int columnIndex = 0; columnIndex < tupleDesc->natts; columnIndex++)
+	{
+		if (columnIndex != 0)
+		{
+			appendStringInfoString(columnDef, ", ");
+		}
+
+		Form_pg_attribute attr = &tupleDesc->attrs[columnIndex];
+		char *attrType = format_type_with_typemod(attr->atttypid, attr->atttypmod);
+
+		appendStringInfo(columnDef, "field_%d %s", columnIndex, attrType);
+	}
+
+	/*
+	 * column definition cannot be empty, so create a dummy column definition for
+	 * queries with no results.
+	 */
+	if (tupleDesc->natts == 0)
+	{
+		appendStringInfo(columnDef, "dummy_field int");
+	}
+
+	StringInfo explainOptions = makeStringInfo();
+	appendStringInfo(explainOptions, "{\"verbose\": %s, \"costs\": %s, \"buffers\": %s, "
+									 "\"timing\": %s, \"summary\": %s, \"format\": \"%s\"}",
+					 CurrentDistributedQueryExplainOptions.verbose ? "true" : "false",
+					 CurrentDistributedQueryExplainOptions.costs ? "true" : "false",
+					 CurrentDistributedQueryExplainOptions.buffers ? "true" : "false",
+					 CurrentDistributedQueryExplainOptions.timing ? "true" : "false",
+					 CurrentDistributedQueryExplainOptions.summary ? "true" : "false",
+					 ExplainFormatStr(CurrentDistributedQueryExplainOptions.format));
+
+	StringInfo wrappedQuery = makeStringInfo();
+	appendStringInfo(wrappedQuery,
+					 "SELECT * FROM worker_save_query_explain_analyze(%s, %s) AS (%s)",
```

Comment:
It's useful useful to explicitly list the columns in the target list in internal queries, because it allows us to add more output columns to the function in a rolling upgrade without breaking things on the coordinator.